dev/user.clj

```diff
@@ -2,7 +2,95 @@
   (:require [cheshire.core :as json]
             [clj-http.client :as http]
             [clj-momo.lib.time :as time]
-            [ctim.schemas.vocabularies :as vocab]))
+            [ctia.init :as init]
+            [ctia.properties :as p]
+            [ctim.schemas.vocabularies :as vocab]
+            [puppetlabs.trapperkeeper.app :as app]
+            [schema.core :as s]))
+
+(set! *warn-on-reflection* true)
+
+;;;;;;;;;;;;;;;;;;;;;;;
+;; Lifecycle management
+;;;;;;;;;;;;;;;;;;;;;;;
+
+;; To avoid losing the current system state, we manually
+;; intern a var in some namespace that is unlikely to be
+;; reloaded.
+
+;; initialize #'ctia.repl.no-reload/system-state
+(s/validate
+  {:app s/Any ;; nil or a running App
+   ;; coordination mechanism for #'serially-alter-app
+   :semaphore java.util.concurrent.Semaphore}
+  (-> (create-ns 'ctia.repl.no-reload)
+      (intern 'system-state)
+      (alter-var-root 
+        #(if (map? %)
+           ;; set root binding exactly once
+           %
+           {:app nil
+            :semaphore (java.util.concurrent.Semaphore. 1)}))))
+
+;; avoids #'locking and private vars to be robust in the presence of code reloading
```

Comment:
Why avoid #'locking?

If the goal is avoiding indefinite locking, `tryLock` from https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html might help.

Semaphores may be considered less intuitive in comparison

Reply:
>  Why avoid #'locking?

I wanted to try and use a coordination mechanism that wasn't tied to any particular bytecode. In this case, if we redefine `serially-alter-app` in the middle of calling it, both definitions will share the same lock.

I didn't fully investigate this, but my understanding is that `locking` generates similar bytecode to Java's `synchronized` methods, and thus two separately compiled `locking` statements cannot coordinate. Does that sound right?

> tryLock from https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html might help.

Ah, thanks! I couldn't find the `Lock` interface (was thinking "flag" for my search terms).

Reply:
> Does that sound right?

It sounds very plausible to me :) good anyway to use the safest possible approach, i.e. Lock is future-proof, as this ns could evolve.